### PR TITLE
P4-2723 initial commit to enforce a new price maintainer role for price maintenance.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
**Changes:**

This change introduces additional role requirements for those users who can add and update journeys prices.

The changes have a dependency on a change to the HMPPS Auth service ([new role](https://github.com/ministryofjustice/hmpps-auth/pull/831)) being rolled out.  At time of writing this has been approved/merged and should be available in DEV.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 plugins {
-  id("uk.gov.justice.hmpps.gradle-spring-boot") version "3.1.3"
+  id("uk.gov.justice.hmpps.gradle-spring-boot") version "3.1.4"
```
Comment (matt-crowson-moj) [inline build.gradle.kts:2]: House keeping change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/SecurityConfiguration.kt [hunk 2]
```diff
@@ -30,6 +31,7 @@ import kotlin.streams.toList
 
 @EnableWebSecurity
 @ConditionalOnWebApplication
+@EnableGlobalMethodSecurity(prePostEnabled = true)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/SecurityConfiguration.kt:34]: This is required to enforce role based access at the class and/or method level in Spring.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/MaintainSupplierPricingController.kt [hunk 3]
```diff
@@ -28,6 +29,7 @@ import javax.validation.constraints.NotNull
   HtmlController.DROP_OFF_ATTRIBUTE,
   HtmlController.DATE_ATTRIBUTE
 )
+@PreAuthorize("hasRole('PECS_MAINTAIN_PRICE')")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/controller/MaintainSupplierPricingController.kt:32]: Using class level as all of the controller methods here required role this particular role enforcement.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/SupplierPricingService.kt [hunk 4]
```diff
@@ -12,6 +13,7 @@ import uk.gov.justice.digital.hmpps.pecs.jpc.price.Supplier
 
 @Service
 @Transactional
+@PreAuthorize("hasRole('PECS_MAINTAIN_PRICE')")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/SupplierPricingService.kt:16]: Same here. Using class level as all of the controller methods here required role this particular role enforcement.

File: src/main/resources/templates/error/403.html [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
```
Comment (matt-crowson-moj) [inline src/main/resources/templates/error/403.html:1]: image

File: src/main/resources/templates/journeys.html [hunk 6]
```diff
@@ -99,10 +99,15 @@ <h1 class="govuk-heading-xl ma0">Journeys for review</h1>
                     <td class="govuk-table__cell" scope="row" th:inline="text">[[${journey.volume}]]</td>
                     <td class="govuk-table__cell" scope="row" th:inline="text">
                         <span th:if="${journey.fromSiteName == null || journey.toSiteName == null}" th:inline="text">Awaiting location</span>
-                        <span th:unless="${journey.fromSiteName == null || journey.toSiteName == null}">
-                            <a th:href="@{'/add-price/' + ${journey.fromNomisAgencyId} + '-' + ${journey.toNomisAgencyId}}"
-                               class="ma0 govuk-link">Add price</a>
-                        </span>
+                        <div sec:authorize="!hasRole('ROLE_PECS_MAINTAIN_PRICE')">
```
Comment (matt-crowson-moj) [inline src/main/resources/templates/journeys.html:102]: You will now see this on the journey for reviews page if you do not have maintain price role:

image

File: src/main/resources/templates/search-journeys-results.html [hunk 7]
```diff
@@ -47,12 +47,19 @@ <h1 class="govuk-heading-xl ma0" th:inline="text">Manage Journey Price Catalogue
                     </a>
                 </td>
                 <td class="govuk-table__cell" scope="row" th:inline="text">[[${journey.toLocationType}]]</td>
-                <td class="govuk-table__cell" scope="row">
-                    <a th:href="@{'/update-price/' + ${journey.fromNomisAgencyId} + '-' + ${journey.toNomisAgencyId}}"
-                       class="ma0 govuk-link" th:inline="text">
-                        £[[${#numbers.formatDecimal(journey.unitPriceInPounds(), 1, 'COMMA', 2, 'POINT')}]]
-                    </a>
-                </td>
+                <div sec:authorize="!hasRole('ROLE_PECS_MAINTAIN_PRICE')">
```
Comment (matt-crowson-moj) [inline src/main/resources/templates/search-journeys-results.html:50]: You will see this on the manage journey prices screen if you do not have the maintain prices role (no link to update price):

image

**Review by Thource — APPROVED** [review]

**Pull request merged**